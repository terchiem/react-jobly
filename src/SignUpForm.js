import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import JoblyApi from './JoblyApi';
import './Form.css';

/** Form component for a user to sign in. 
 * 
 * Props:
 *  setToken -> App component state setter for token
 * 
 * State:
 *  formData -> state for each input on the form
*/

function SignUpForm({ setToken }) {
  const history = useHistory();

  const [formData, setFormData] = useState({
    username: "",
    password: "",
    first_name: "",
    last_name: "",
    email: "",
    errors: []
  });

  /** Updates state on input change */
  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData(data => ({...data, [name]: value}));
  }

  /** Creates a new user and stores the token in localStorage */
  async function signUp(evt) {
    evt.preventDefault();
    
    try {
      const token = await JoblyApi.signUp(formData);
      setToken(token);
      history.push("/jobs");  // redirect to '/jobs'
    } catch (errors) {
      // fill formData errors array to display
      setFormData(data => ({...data, errors}));
    }
  }

  /** Renders form errors if present */
  function renderErrors() {
    return (formData.errors.length ?
      <ul className="Form-errors">
        {formData.errors.map((message, i) => (
          <li key={i}>{message}</li>
        ))}
      </ul> : null
    )
  }

  return (
    <form className="Form" onSubmit={signUp}>
      <div className="Form-group">
        <label className="Form-label" htmlFor="username">Username</label>
        <input
          className="Form-input" 
          id="username" 
          name="username" 
          value={formData.username} 
          onChange={handleChange} 
        />
      </div>
      <div className="Form-group">
        <label className="Form-label" htmlFor="password">Password</label>
        <input
          className="Form-input" 
          id="password" 
          name="password" 
          type="password"
          value={formData.password} 
          onChange={handleChange} 
        />
      </div>
      <div className="Form-group">
        <label className="Form-label" htmlFor="first_name">First name</label>
        <input
          className="Form-input" 
          id="first_name" 
          name="first_name" 
          value={formData.first_name} 
          onChange={handleChange} 
        />
      </div>
      <div className="Form-group">
        <label className="Form-label" htmlFor="last_name">Last name</label>
        <input
          className="Form-input" 
          id="last_name" 
          name="last_name" 
          value={formData.last_name} 
          onChange={handleChange} 
        />
      </div>
      <div className="Form-group">
        <label className="Form-label" htmlFor="email">Email</label>
        <input
          className="Form-input" 
          id="email" 
          name="email" 
          type="email"
          value={formData.email} 
          onChange={handleChange} 
        />
      </div>
      <button className="Form-submit">Submit</button>

      { renderErrors() }
    </form>
  )
}

export default SignUpForm;