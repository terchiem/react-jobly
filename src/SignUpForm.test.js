import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import SignUpForm from './SignUpForm';

test('renders without crashing', () => {
  render(
    <MemoryRouter>
      <SignUpForm />
    </MemoryRouter>
  );
});

test('matches snapshot', () => {
  const { asFragment } = render(
    <MemoryRouter>
      <SignUpForm />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});