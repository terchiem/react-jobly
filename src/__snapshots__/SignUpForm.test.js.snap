// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="Form"
  >
    <div
      class="Form-group"
    >
      <label
        class="Form-label"
        for="username"
      >
        Username
      </label>
      <input
        class="Form-input"
        id="username"
        name="username"
        value=""
      />
    </div>
    <div
      class="Form-group"
    >
      <label
        class="Form-label"
        for="password"
      >
        Password
      </label>
      <input
        class="Form-input"
        id="password"
        name="password"
        type="password"
        value=""
      />
    </div>
    <div
      class="Form-group"
    >
      <label
        class="Form-label"
        for="first_name"
      >
        First name
      </label>
      <input
        class="Form-input"
        id="first_name"
        name="first_name"
        value=""
      />
    </div>
    <div
      class="Form-group"
    >
      <label
        class="Form-label"
        for="last_name"
      >
        Last name
      </label>
      <input
        class="Form-input"
        id="last_name"
        name="last_name"
        value=""
      />
    </div>
    <div
      class="Form-group"
    >
      <label
        class="Form-label"
        for="email"
      >
        Email
      </label>
      <input
        class="Form-input"
        id="email"
        name="email"
        type="email"
        value=""
      />
    </div>
    <button
      class="Form-submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
